File: packages/prompt-core/src/mcp-registry.ts
import type {
  McpServerConfig,
  McpUrlServerConfig,
  McpStdioServerConfig,
} from "./mcp";
import { interpolateEnvInObject } from "./mcp";

/**
 * Minimal interface an MCP client must satisfy. Every SDK's MCP client
 * (AI SDK v4/v5, Mastra, etc.) exposes `tools()` returning a map keyed by
 * tool name — that's all the registry needs. The actual client import is
 * deferred to the executor package via the `createClient` factory so
 * `prompt-core` stays free of SDK dependencies.
 */
export interface McpClient<TTool> {
  tools(): Promise<Record<string, TTool>>;
}

/**
 * Factory that creates a client given the raw server config. The factory
 * is responsible for picking the right transport (SSE/stdio) and importing
 * the SDK's MCP client dynamically as a peer dependency.
 */
export type McpClientFactory<TTool> = (
  config: McpUrlServerConfig | McpStdioServerConfig
) => Promise<McpClient<TTool>>;

function isUrlConfig(cfg: McpServerConfig): cfg is McpUrlServerConfig {
  return (cfg as McpUrlServerConfig).url !== undefined;
}

function isStdioConfig(cfg: McpServerConfig): cfg is McpStdioServerConfig {
  return (cfg as McpStdioServerConfig).command !== undefined;
}

/**
 * Shared MCP server registry. Resolves `mcp://server/tool` URIs (including
 * the `mcp://server/*` wildcard) to SDK-native tool values. Connection +
 * tool listing is cached per server; client construction is deferred to
 * the per-SDK `McpClientFactory` so this package has zero MCP runtime deps.
 */
export class McpServerRegistry<TTool = unknown> {
  private readonly servers = new Map<string, McpServerConfig>();
  private readonly clients = new Map<string, Promise<McpClient<TTool>>>();
  private readonly toolsCache = new Map<string, Record<string, TTool>>();

  constructor(private readonly factory: McpClientFactory<TTool>) {}

  register(name: string, config: McpServerConfig): this {
    this.servers.set(name, config);
    return this;
  }

  registerServers(servers: Record<string, McpServerConfig>): this {
    for (const [name, config] of Object.entries(servers)) {
      this.register(name, config);
    }
    return this;
  }

  has(name: string): boolean {
    return this.servers.has(name);
  }

  getConfig(name: string): McpServerConfig | undefined {
    return this.servers.get(name);
  }

  private getConfigOrThrow(name: string): McpServerConfig {
    const cfg = this.servers.get(name);
    if (!cfg) {
      throw new Error(
        `MCP server '${name}' not registered. Available servers: ${Array.from(
          this.servers.keys()
        ).join(", ")}`
      );
    }
    return cfg;
  }

  private async createClient(serverName: string): Promise<McpClient<TTool>> {
    const rawCfg = this.getConfigOrThrow(serverName);
    const cfg = interpolateEnvInObject(rawCfg);

    if (isUrlConfig(cfg) || isStdioConfig(cfg)) {
      return this.factory(cfg);
    }

    throw new Error("Invalid MCP server config: expected 'url' or 'command'");
  }

  async getClient(serverName: string): Promise<McpClient<TTool>> {
    const existing = this.clients.get(serverName);
    if (existing) return existing;
    const created = this.createClient(serverName);
    this.clients.set(serverName, created);
    created.catch((err: unknown) => {
      console.error(
        `[McpServerRegistry] Failed to connect to MCP server '${serverName}':`,
        err
      );
      this.clients.delete(serverName);
    });
    return created;
  }

  async getTool(serverName: string, toolName: string): Promise<TTool> {
    const existingTools = this.toolsCache.get(serverName);
    if (existingTools && existingTools[toolName]) {
      return existingTools[toolName];
    }

    let allTools: Record<string, TTool>;
    try {
      const client = await this.getClient(serverName);
      allTools = await client.tools();
      this.toolsCache.set(serverName, allTools);
    } catch (err) {
      this.toolsCache.delete(serverName);
      throw err;
    }

    const tool = allTools[toolName];
    if (!tool) {
      throw new Error(
        `MCP tool not found: ${serverName}/${toolName}. Available tools: ${Object.keys(
          allTools
        ).join(", ")}`
      );
    }
    return tool;
  }

  async getAllTools(serverName: string): Promise<Record<string, TTool>> {
    const existingTools = this.toolsCache.get(serverName);
    if (existingTools) {
      return existingTools;
    }

    try {
      const client = await this.getClient(serverName);
      const allTools = await client.tools();
      this.toolsCache.set(serverName, allTools);
      return allTools;
    } catch (err) {
      this.toolsCache.delete(serverName);
      throw err;
    }
  }
}

File: packages/prompt-core-python/src/agentmark/prompt_core/mcp_registry.py
"""Shared MCP server registry.

Mirrors the TypeScript `McpServerRegistry` in
`@agentmark-ai/prompt-core/mcp-registry`. Generic over the SDK's native
tool type via TypeVar. Client construction is deferred to a factory
callable so this module stays free of any SDK dependency.
"""

from __future__ import annotations

import logging
from collections.abc import Awaitable, Callable
from typing import Any, Generic, Protocol, TypeVar, runtime_checkable

from .mcp import (
    McpServerConfig,
    McpStdioServerConfig,
    McpUrlServerConfig,
    interpolate_env_in_object,
)

log = logging.getLogger(__name__)

TTool = TypeVar("TTool")


@runtime_checkable
class McpClient(Protocol[TTool]):
    """Minimal contract an MCP client must satisfy."""

    async def tools(self) -> dict[str, TTool]: ...


# Factory signature: takes interpolated config, returns a connected client.
McpClientFactory = Callable[[McpServerConfig], Awaitable[McpClient[TTool]]]


def _is_url_config(cfg: McpServerConfig) -> bool:
    return "url" in cfg and cfg.get("url") is not None


def _is_stdio_config(cfg: McpServerConfig) -> bool:
    return "command" in cfg and cfg.get("command") is not None


class McpServerRegistry(Generic[TTool]):
    """Resolves `mcp://server/tool` URIs (including wildcards) to SDK-native
    tool values. Connection + tool listing is cached per server."""

    def __init__(self, factory: McpClientFactory[TTool]) -> None:
        self._factory = factory
        self._servers: dict[str, McpServerConfig] = {}
        self._clients: dict[str, McpClient[TTool]] = {}
        self._tools_cache: dict[str, dict[str, TTool]] = {}

    def register(self, name: str, config: McpServerConfig) -> "McpServerRegistry[TTool]":
        self._servers[name] = config
        return self

    def register_servers(
        self, servers: dict[str, McpServerConfig]
    ) -> "McpServerRegistry[TTool]":
        for name, config in servers.items():
            self.register(name, config)
        return self

    def has(self, name: str) -> bool:
        return name in self._servers

    def get_config(self, name: str) -> McpServerConfig | None:
        return self._servers.get(name)

    def _get_config_or_raise(self, name: str) -> McpServerConfig:
        cfg = self._servers.get(name)
        if cfg is None:
            raise ValueError(
                f"MCP server '{name}' not registered. "
                f"Available servers: {', '.join(self._servers.keys()) or '(none)'}"
            )
        return cfg

    async def _create_client(self, server_name: str) -> McpClient[TTool]:
        raw_cfg = self._get_config_or_raise(server_name)
        cfg = interpolate_env_in_object(raw_cfg)
        if _is_url_config(cfg) or _is_stdio_config(cfg):
            return await self._factory(cfg)
        raise ValueError("Invalid MCP server config: expected 'url' or 'command'")

    async def get_client(self, server_name: str) -> McpClient[TTool]:
        existing = self._clients.get(server_name)
        if existing is not None:
            return existing
        try:
            created = await self._create_client(server_name)
            self._clients[server_name] = created
            return created
        except Exception as err:
            log.error(
                "[McpServerRegistry] Failed to connect to MCP server '%s': %s",
                server_name,
                err,
            )
            raise

    async def get_tool(self, server_name: str, tool_name: str) -> TTool:
        cached = self._tools_cache.get(server_name)
        if cached is not None and tool_name in cached:
            return cached[tool_name]

        try:
            client = await self.get_client(server_name)
            all_tools = await client.tools()
            self._tools_cache[server_name] = all_tools
        except Exception:
            self._tools_cache.pop(server_name, None)
            raise

        tool = all_tools.get(tool_name)
        if tool is None:
            raise ValueError(
                f"MCP tool not found: {server_name}/{tool_name}. "
                f"Available tools: {', '.join(all_tools.keys())}"
            )
        return tool

    async def get_all_tools(self, server_name: str) -> dict[str, TTool]:
        cached = self._tools_cache.get(server_name)
        if cached is not None:
            return cached
        try:
            client = await self.get_client(server_name)
            all_tools = await client.tools()
            self._tools_cache[server_name] = all_tools
            return all_tools
        except Exception:
            self._tools_cache.pop(server_name, None)
            raise
